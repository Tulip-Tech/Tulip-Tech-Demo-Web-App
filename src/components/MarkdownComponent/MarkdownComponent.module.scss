@use 'src/styles/variables';
@use 'src/styles/theme';

.markdown {
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 0 0 1em;
    font-weight: bold;
    line-height: 1.33;
  }

  h1 {
    padding-bottom: 0.3em;
    font-size: 2em;
    border-bottom: 1px solid rgba(variables.$white, 0.12);
  }

  h2 {
    padding-bottom: 0.3em;
    font-size: 1.5em;
    border-bottom: 1px solid rgba(variables.$white, 0.12);
  }

  h3 {
    font-size: 1.25em;
  }

  h4 {
    font-size: 1em;
  }

  h5 {
    font-size: 0.875em;
  }

  h6 {
    font-size: 0.85em;
  }

  hr {
    box-sizing: content-box;
    height: 0.25em;
    margin: 24px 0;
    padding: 0;
    overflow: hidden;
    background: transparent;
    background-color: rgba(variables.$white, 0.12);
    border: 0;
  }

  img {
    box-sizing: content-box;
    max-width: 100%;
    background-color: theme.$secondary-color;
  }

  a,
  a:visited {
    color: theme.$link-color;
    text-decoration: none;

    &:hover {
      text-decoration: underline;
    }
  }

  ul,
  ol,
  p {
    margin: 1.2em 0;
  }
}
