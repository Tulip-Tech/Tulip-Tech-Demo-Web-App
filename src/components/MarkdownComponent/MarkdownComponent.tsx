import React, { useMemo } from 'react';
import DOMPurify from 'dompurify';
import { marked } from 'marked';
import classNames from 'classnames';

import styles from './MarkdownComponent.module.scss';

const renderer = {
  link(href: string, title: string, text: string) {
    const externalLink = /^(https?|www\.|\/\/)/.test(href || '');
    const targetAttr = externalLink ? 'target="_blank"' : undefined;
    const relAttr = externalLink ? 'rel="noopener"' : undefined;
    const titleAttr = title ? `title="${title}"` : undefined;
    const attributes = [targetAttr, relAttr, titleAttr].filter(Boolean);

    return `<a href="${href}" ${attributes.join(' ')}>${text}</a>`;
  },
};

marked.use({ renderer });

type Props = {
  markdownString: string;
  className?: string;
  inline?: boolean;
};

const MarkdownComponent: React.FC<Props> = ({ markdownString, className, inline = false }) => {
  const sanitizedHTMLString = useMemo(() => {
    const parseDelegate = inline ? marked.parseInline : marked.parse;
    const dirtyHTMLString = parseDelegate(markdownString);

    return DOMPurify.sanitize(dirtyHTMLString, { ADD_ATTR: ['target'] });
  }, [inline, markdownString]);

  return <div className={classNames(styles.markdown, className)} dangerouslySetInnerHTML={{ __html: sanitizedHTMLString }} />;
};

export default MarkdownComponent;
