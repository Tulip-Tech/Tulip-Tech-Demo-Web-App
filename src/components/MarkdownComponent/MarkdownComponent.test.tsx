import React from 'react';
import { render } from '@testing-library/react';

import MarkdownComponent from './MarkdownComponent';

describe('<MarkdownComponent>', () => {
  test('renders and matches snapshot', () => {
    const markdownString = 'This is a test markdown string with a [testlink](https://www.example.com)';
    const { container } = render(<MarkdownComponent markdownString={markdownString} />);

    expect(container).toMatchSnapshot();
  });

  test('should create a save and valid link element', () => {
    const markdownString = 'This is a test markdown string with a [testlink](https://www.example.com)';
    const { getByText } = render(<MarkdownComponent markdownString={markdownString} />);

    expect(getByText('testlink')).toHaveAttribute('href', 'https://www.example.com');
    expect(getByText('testlink')).toHaveAttribute('rel', 'noopener');
    expect(getByText('testlink')).toHaveAttribute('target', '_blank');
  });

  test('should create a valid link element for local pages', () => {
    const markdownString = 'This is a test markdown string with a [testlink](/page)';
    const { getByText } = render(<MarkdownComponent markdownString={markdownString} />);

    expect(getByText('testlink')).toHaveAttribute('href', '/page');
    expect(getByText('testlink')).not.toHaveAttribute('rel', 'noopener');
    expect(getByText('testlink')).not.toHaveAttribute('target', '_blank');
  });
});
