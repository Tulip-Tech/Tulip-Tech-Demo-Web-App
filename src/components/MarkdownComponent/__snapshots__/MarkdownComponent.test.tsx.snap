// Vitest Snapshot v1

exports[`<MarkdownComponent> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_markdown_e219e2"
  >
    <p>
      This is a test markdown string with a 
      <a
        href="https://www.example.com"
        rel="noopener"
        target="_blank"
      >
        testlink
      </a>
    </p>
    

  </div>
</div>
`;
